Cadastro autor

savitoh opened the pull request:
Precisa aperfeiçoar mecanismo de validação do e-mail.

savitoh committed:
Adição LIB Lombok; Start Flyway Scripts

savitoh committed:
Conclusao Cadastro Autor

asouza commented on src/main/java/com/savitoh/casacodigoapi/dto/AutorDto.java:
```diff
@@ -0,0 +1,40 @@
+
+    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "dd/MM/yyyy HH:mm:ss")
+    @JsonProperty(access = JsonProperty.Access.READ_ONLY)
+    private LocalDateTime createdAt;
```
A data de criação do autor precisa ser o instante atual. Se você deixa essa informação do lado de fora da classe Autor, quer dizer que o mesmo autor pode ser criado com uma data diferente do instante atual,

savitoh replied:
Acredito que o problema estar em permitir, na Classe Autor, setar esse parâmetro.

asouza replied:
Sim. Mas eu confesso que não entendi pq o AutorDto precisa de um atributo relativo a data de criação.

asouza commented on src/main/java/com/savitoh/casacodigoapi/dto/AutorDto.java:
```diff
@@ -0,0 +1,40 @@
+public class AutorDto implements Serializable {
+
+
+    private Integer codigo;
```
Nesse momento do código, eu confesso que não captei a necessidade do atributo do ```codigo```

asouza commented on src/main/java/com/savitoh/casacodigoapi/dto/AutorDto.java:
```diff
@@ -0,0 +1,40 @@
+@AllArgsConstructor
+@NoArgsConstructor
+@Builder
+public class AutorDto implements Serializable {
```
*A complicação do nosso é código é proporcional a complicação da nossa feature. Quanto mais simples, melhor.*

Pq essa classe tem construtor com argumentos, getters e setters? O spring conseguiria montar o objeto em função do formulário através do construtor ou através dos métodos de acesso. Não precisa dos dois.

savitoh replied:
Mal Hábito.

asouza commented on src/main/java/com/savitoh/casacodigoapi/dto/AutorDto.java:
```diff
@@ -0,0 +1,40 @@
+@AllArgsConstructor
+@NoArgsConstructor
+@Builder
+public class AutorDto implements Serializable {
```
Ainda em relação a complicação. 

O lombok é mega complicado. Mesmo as pessoas que trabalham tem dificuldade em explicar o mecanismo de funcionamento da biblioteca. Como que ele mexe na estrutura da sua classe para gerar os métodos de acesso, construtores etc? Imagina que nunca viu isso. A pessoa tem que ter um belo background para entender e não deixar essa pergunta martelando na mente enquanto produz... Ou seja, ela vai produzir menos por um tempo pq parte do cerebro dela está pensando em como funciona o lombok. 

Por fim, será que usá-lo é realmente tão bom? Qual a diferença de você mandar um @Getter e @Setter em vez de escrever? Imagine que o código é gerado pelo editor e que número de linhas de código não é uma métrica confiável para qualquer critério de qualidade.

asouza commented on src/main/java/com/savitoh/casacodigoapi/model/Autor.java:
```diff
@@ -0,0 +1,61 @@
+@AllArgsConstructor
+@NoArgsConstructor
+@Builder
+public class Autor implements Serializable {
```
*A complicação do nosso é código é proporcional a complicação da nossa feature. Quanto mais simples, melhor.*
+ 
*Usamos tudo que conhecemos que está pronto. Só fazemos código do zero se for estritamente necessário.*



Tem quase mais anotação do que linha de código. Algumas perguntas?
* Pq usar o @Table para gerar a tabela com o mesmo nome que já seria gerada? 
* Pq precisa de um construtor sem argumentos?
* Pq precisa de um construtor que recebe o id e o instante de criação?
* Pq precisa de um Builder quando tem um construtor simples?
* Pq implementa Serializable se não precisa?
* Pq o @Column especifica o nome da coluna se o default é que o nome da coluna seja o mesmo do atributo?

Todas essas perguntas podem ser feitas por quem for manter esse código. E se ela quiser saber de tudo isso, vai ser menos eficiente naquele intervalo de tempo.

asouza replied:
sobre a sequence em si, esse já é o modo identity do pg, se não me engano. E tudo bem se não for e você quiser configurar :).

asouza replied:
Especificamente sobre o id no construtor. O id aí só existe por conta de uma determinação do banco de dados. Não tem nenhuma relação com o domínio. Se não tem relação com o domínio, não existe necessidade de construir o objeto passando ele :).

asouza commented on src/main/java/com/savitoh/casacodigoapi/model/Autor.java:
```diff
@@ -0,0 +1,61 @@
+    private LocalDateTime createdAt;
+
+    @Override
+    public boolean equals(Object o) {
```
Aqui tem a ver com regra de negócio. 

Igualdade dos objetos é definida pela pessoa que pediu a funcionalidade. Além da gente não ter especificado isso, a igualdade não pode ser pelo ```id```. Toda igualdade e hashcode deve ser feita pelas informações que são naturais do objeto. Caso contrário você fica dependente do banco para realizar qualquer comparação entre os dois. A igualdade de dois objetos sempre é feita pelos atributos que são preenchidos no construtor. Pq ali é o mínimo que você precisa para montar a instância.

asouza replied:
E como a gente não precisa do equals e hashcode ainda, não precisamos criar os métodos. Todo método público que você cria devia ser chamado por alguém.

asouza commented on src/main/java/com/savitoh/casacodigoapi/resource/AutorResource.java:
```diff
@@ -0,0 +1,32 @@
+@RequestMapping("/v1/autor")
+public class AutorResource {
+
+    private final AutorService autorService;
```
*A complicação do nosso é código é proporcional a complicação da nossa feature. Quanto mais simples, melhor.*

O service aqui chama o repositório que chama o save.  Agora esqueça o que você leu sobre service e qualquer padrão que você use no seu dia a dia de trabalho. A pergunta que fica é: logicamente falando, pq eu preciso de um service? Pq não chamo o repository direto? Mais ainda. Pq eu preciso de um repository e não chamo o EntityManager direto aqui? Pq seria ruim?

asouza commented on src/main/java/com/savitoh/casacodigoapi/resource/AutorResource.java:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    @PostMapping
+    public ResponseEntity<AutorDto> criaAutor(@Valid @RequestBody AutorDto autorDto) {
```
*A complicação do nosso é código é proporcional a complicação da nossa feature. Quanto mais simples, melhor.* 
+
*Avaliamos de forma lógica o nível de complicação de cada trecho de código. Aqui não tem espaço para feeling. Usamos a teoria da carga cognitiva e medimos o nível de pontos de entendimentos necessário por trecho.*

Aqui você usa 4 abstrações para salvar um autor. A classe Autor, o mapper, o service + o DTO. Cada abstração aumenta um ponto no nível de dificuldade do código. Pq isso é melhor que ```manager.persit(new Autor(dto.getNome(),dto.getDescricao()))``` ?

asouza replied:
Como eu falei antes também, o DTO adiciona uma complexidade a mais também. Só que eu sugiro o uso por conta que os parâmetros da web realmente tendem a ficar diferentes das propriedades do objeto em algum momento. Temos formulários com edições parciais etc... E aí começa o jogo da balança.

asouza commented on src/main/java/com/savitoh/casacodigoapi/mapper/AutorMapper.java:
```diff
@@ -0,0 +1,35 @@
+
+import java.util.Objects;
+
+public final class AutorMapper {
```
*A complicação do nosso é código é proporcional a complicação da nossa feature. Quanto mais simples, melhor.*

Pq precisamos de um mapper? Pq não usar o construtor do autor direto no controller? Ou pq não criar um método no DTO que retorna um novo autor? Se você cria esse método lá você deixa o estado perto do comportamento e tende a facilitar o entendimento. Vamos tentar esquecer nossos padrões :).

asouza commented on src/main/java/com/savitoh/casacodigoapi/mapper/AutorMapper.java:
```diff
@@ -0,0 +1,35 @@
+            throw new IllegalArgumentException("Argumento Não Pode Ser Nullo");
+        }
+        return Autor.builder()
+                .autorId(autorDto.getCodigo())
```
*Não retornamos nulo dentro das regras da aplicação. Pense que seu computador vai explodir.*

Se esse código fosse ser mantido, teríamos que dar um jeito nesse getCodigo. Pq, neste momento, ele permite retornar nulo. Quando retornamos nulo, tomamos NPE.

asouza commented on src/main/java/com/savitoh/casacodigoapi/service/GenericService.java:
```diff
@@ -0,0 +1,16 @@
+import java.util.List;
+import java.util.Optional;
+
+public interface GenericService <T extends Serializable, Id extends Serializable> {
```
Sugiro apagar por tudo que foi explicado antes.

asouza commented on src/main/java/com/savitoh/casacodigoapi/service/AutorService.java:
```diff
@@ -0,0 +1,44 @@
+
+@Service
+@Transactional(readOnly = false)
+public class AutorService implements GenericService<Autor, Integer> {
```
Sugiro apagar por conta do que foi explicado antes.

asouza commented on src/main/java/com/savitoh/casacodigoapi/service/AutorService.java:
```diff
@@ -0,0 +1,44 @@
+import java.util.Optional;
+
+@Service
+@Transactional(readOnly = false)
```
*Usamos tudo que conhecemos que está pronto. Só fazemos código do zero se for estritamente necessário.*

Mesmo que essa classe fosse ficar... O readOnly é false por default. Tudo que você escreve vai querer ser interpretado por alguém.. E quanto mais coisa precisa ser interpretada, mais complicado é o código.

savitoh committed:
Refactor para impedir campo dataCriacao(Autor) ser alterado;
Cria metodo para transormar dto em entidade (AutorDto)
Remove Lombok Lib;
Remove Mapper, AutorService, ServiceGeneric;

savitoh committed:
Remove implementação Serializable

savitoh reviewed (commented):
Alterações feitas. Achei apenas drástico a remoção da camada Service.

asouza commented on src/main/java/com/savitoh/casacodigoapi/exception/ApiExceptionGlobalHandler.java:
```diff
@@ -0,0 +1,44 @@
+
+@Order(Ordered.HIGHEST_PRECEDENCE)
+@RestControllerAdvice
+public class ApiExceptionGlobalHandler extends ResponseEntityExceptionHandler {
```
É bom ter essa classe na manga mesmo, basicamente ela é repetida em todo santo projeto.

asouza commented on src/main/java/com/savitoh/casacodigoapi/model/Autor.java:
```diff
@@ -1,61 +1,84 @@
 
     @CreationTimestamp
     @Column(name = "created_at", columnDefinition = "TIMESTAMP")
     private LocalDateTime createdAt;
```
Pq não faz ```createdAt = LocalDateTime.now()```? Digo, tem alguma vantagem real em delegar esse comportamento para o hibernate? Do jeito que está seu objeto está nascendo em estado inconsistente.

asouza commented on src/main/java/com/savitoh/casacodigoapi/model/Autor.java:
```diff
@@ -1,61 +1,84 @@
-        if (o == null || getClass() != o.getClass()) return false;
-        Autor autor = (Autor) o;
-        return Objects.equals(autorId, autor.autorId);
+    public Autor() {
```
Precisamos deste construtor aqui?

asouza commented on src/main/java/com/savitoh/casacodigoapi/resource/AutorResource.java:
```diff
@@ -17,16 +16,16 @@
-        return new ResponseEntity<>(AutorMapper.entityToDto(autorSaved), HttpStatus.CREATED);
+        Autor autor = autorDto.transformaParaEntity();
+        Autor autorSalvo = autorRepository.save(autor);
+        return new ResponseEntity<>(autorDto.transformaEntityParaDto(autorSalvo), HttpStatus.CREATED);
```
esse método que transforma a entidade para um DTO tende a ser estático.. já que ele usa apenas as informações que vem via parâmetro. 

Aqui tem um detalhe: você o mesmo DTO para a entrada e a saída do objeto autor. A chance é grande de você mudar alguma característica nesse dto e, sem querer, quebrar um dos dois endpoints. Os objetos de entrada e saída da borda mais externa tendem a ser super específicos e o melhor é deixar separado mesmo.. Digo, eles tem ligação direta com o endpoint.

savitoh replied:
Sim sim. Um payload para resquest e outro para response?

asouza replied:
Do meu ponto de vista sim. A coincidência de mesmos atributos faz com que a gente enxergue que a mesma classe pode ser usada para duas situações diferentes, mas tende a ser uma cilada.

savitoh replied:
Concordo. Mas um ponto, no momento, não seria um esforço além da necessidade?

asouza replied:
Boa pergunta. Acho que sim, mas essa é uma prática que eu sigo desde o ínicio, mesmo sabendo que ela é uma otimização prematura. Ótimo questionamento. 

Por outro lado, a spec do cartão dizia que era apenas para retornar 200 :P.

asouza reviewed (approved):
Ainda existem pontos de melhoria, mas, na minha opinião, nada que não possa ser trabalhado durante os próximos prs.
